### DISCUSSION TITLE
No Sound in Playback

### DISCUSSION THREAD

**Discussion by gregoireLem** [HaishinKit/HaishinKit.swift#1398] [Q&A] [answered]
First thanks a lot for the repo 🫡

My problem is I don't have audio playback when playing a video via SRT using MTHKView.
I tried with a video with aac & mp3.

I get the following error in the console when starting playing :

```
AVAEInternal.h:76    required condition is false: [AVAudioNodeImpl.h:296:AVAE_CheckNodeHasEngine: (_engine != nil)]
[Warn] [com.haishinkit.HaishinKit] [MediaLink.swift:29] isPaused > required condition is false: _engine != nil
```

Here is how I initiate the AVAudioSession in app delegate.
Also audio background mode is enabled in app capabilities.

```
let session = AVAudioSession.sharedInstance()
        do {
            try session.setCategory(.playback, mode: .moviePlayback)
            try session.setActive(true)
        } catch {
            print("Setting category to AVAudioSessionCategoryPlayback failed.")
        }
        return true
```
 
 I tried to let audioSettings to default, as well as initiating the audio setting with my video source spec : 
 
```
        let audioCodecSettings = AudioCodecSettings(
            bitRate: 130000,
            sampleRate: 44100,
            channels: 1,
            downmix: false,
            channelMap: nil
        )
        
       srtStream.audioSettings =  audioCodecSettings
```

Thanks for your help

**Comment by shogo4405** [comment]
1. MP3 format is not supported at all.
1. Specifying audioSettings is unnecessary during playback.
1. The error message occurs when data is not received from the server at all, such as during ingest. Can you provide the application name (e.g., OBS) that is ingesting or the server name or service name used on the server side?

Thanks.

**Reply by gregoireLem** [reply]
Thanks for your response.
I use a local server to test with ffmpeg (the video sound is aac) : 

ffmpeg -stream_loop -1 -re -i input.mp4 -c copy -f mpegts 'srt://0.0.0.0:9998?mode=listener'

And when playing it via : 

ffplay 'srt://localhost:9998?mode=caller'

I do have the sound.

However with HaishinKit I only have video, that's why I thought it's a problem on how I initiate HaishinKit or the audio in my app.

Thanks again

**Comment by gregoireLem** [comment]
Also I do have sound using player pro app : https://apps.apple.com/fr/app/haivision-play-pro/id1482925169

**Comment by gregoireLem** [comment]
To help debug : in **IOAudioUnit.swift** I printed the sampled buffer received : 
   
```
    func append(_ sampleBuffer: CMSampleBuffer, channel: UInt8 = 0) {
        switch sampleBuffer.formatDescription?.audioStreamBasicDescription?.mFormatID {
        case kAudioFormatLinearPCM:
            resampler.append(sampleBuffer.muted(muted))
        default:
            print(sampleBuffer)
            codec.append(sampleBuffer)
        }
    }
```
``` 
    CMSampleBuffer 0x103f15900 retainCount: 3 allocator: 0x1e9a4a918
	invalid = NO
	dataReady = YES
	makeDataReadyCallback = 0x0
	makeDataReadyRefcon = 0x0
	formatDescription = <CMAudioFormatDescription 0x60000330c3c0 [0x1e9a4a918]> {
	mediaType:'soun' 
	mediaSubType:'aac ' 
	mediaSpecific: {
		ASBD: {
			mSampleRate: 44100.000000 
			mFormatID: 'aac ' 
			mFormatFlags: 0x2 
			mBytesPerPacket: 0 
			mFramesPerPacket: 1024 
			mBytesPerFrame: 0 
			mChannelsPerFrame: 2 
			mBitsPerChannel: 0 	} 
		cookie: {(null)} 
		ACL: {(null)}
		FormatList Array: {
			Index: 0 
			ChannelLayoutTag: 0x650002 
			ASBD: {
			mSampleRate: 44100.000000 
			mFormatID: 'aac ' 
			mFormatFlags: 0x2 
			mBytesPerPacket: 0 
			mFramesPerPacket: 1024 
			mBytesPerFrame: 0 
			mChannelsPerFrame: 2 
			mBitsPerChannel: 0 	}} 
	} 
	extensions: {(null)}
}
```

**Reply by shogo4405** [reply]
I fixed it. Please take a look.
https://github.com/shogo4405/HaishinKit.swift/pull/1401

**Comment by gregoireLem** [comment]
Hey thanks for checking the issue, but unfortunately I still have no sound
(tried on simulator and device with the video I gave you).
What can I do to help the debug process ?

<img width="786" alt="Screenshot 2024-04-06 at 11 25 07" src="https://github.com/shogo4405/HaishinKit.swift/assets/31652880/ed159336-e11d-44de-8079-633b9445fa2a">

**Reply by shogo4405** [reply]
Certainly, it didn't work in iOS testing. I was testing it on macOS. When I made the following changes to the iOS sample, it worked. Please verify.
https://github.com/shogo4405/HaishinKit.swift/tree/main/Examples
```
// If you set the "mode" parameter, stereo capture is not possible, so it is left unspecified.
// try session.setCategory(.playAndRecord, options: [.defaultToSpeaker, .allowBluetooth])
            
try session.setCategory(.playback)
try session.setActive(true)
```

**Comment by gregoireLem** [comment]
Ok so i still have no sound when the video plays.
But when I move the app in the background, wait 5 secs, come back to the
app, the video stops and the sound starts. it's like I can't have both at
the same time

**Reply by shogo4405** [reply, marked as answer]
I have confirmed that it works with https://github.com/shogo4405/HaishinKit.swift/tree/main/Examples.
Is it exhibiting similar symptoms with this code as well?

If it is not working, please report following this template:
https://github.com/shogo4405/HaishinKit.swift/issues/new?assignees=&labels=&projects=&template=bug_report.yml

**Comment by shogo4405** [comment]
Close as fixed https://github.com/shogo4405/HaishinKit.swift/pull/1401
